### PR TITLE
Improve ci data management

### PR DESCRIPTION
**Pull request by mdiazmel**
This PR propose a better organization of the data in the CI agents. 

- [X] A fixture at the moment of launch the test setup the folders to find the input data and the working dir.
- [X] An unique $TMPDIR is automatically generated by the tests to save outputs (by using [tmp_path](https://docs.pytest.org/en/6.2.x/tmpdir.html#base-temporary-directory))
- [X] Remove mac tests for build, install and instantiate.
- [X] Activate non-regression tests for convertors and iotools in each PR.

### CODE REVIEW

**Comment by mdiazmel** [comment]
`CreateMergeFile` test is disable and it will enable in a different PR (results are not the same wrt reference data in CI agents, probably there are environment incompatibilities).

File: test/nonregression/iotools/test_run_converters.py [hunk 1]
```diff
@@ -18,140 +19,118 @@
...
-    clean_folder(join(root, "out", "bids"), recreate=True)
+    base_dir = Path(cmdopt["input"])
+    input_dir = base_dir / name / "in"
+    ref_dir = base_dir / name / "ref"
+    tmp_out_dir = tmp_path / name / "out"
+    tmp_out_dir.mkdir(parents=True)
+    dataset_directory = input_dir / "unorganized"
+    bids_directory = tmp_out_dir / "bids"
+
+    if name == "Nifd2Bids":
```
[the remark this answers is not in the document]
Comment (ghisvail) [reply test/nonregression/iotools/test_run_converters.py:52]: I agree with Omar. The nice thing with multiple functions is that they provide a contextual boundary for what is covered by the actual test. Before this, it was very clear from the layout of the code that we had one test for each dataset.

Now we have a single large function with multiple `if` branches (which is a code smell in itself), and it becomes more cognitively challenging to maintain _and_ extend the test. If the aim was to factor out the necessary boilerplate to access the files, then using `pytest` setup and teardown mechanisms should be more appropriate.

[the remark this answers is not in the document]
Comment (mdiazmel) [reply test/nonregression/iotools/test_run_converters.py:52]: > why is this refactoring necessary?

Indeed, what I propose to do here is to use [pytest fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) (see line 28) to run all the tests for the converters (same for iotools and for instance tests). Each fixture, previously declared,  has a different context and the tests are completely isolated. 

  > It seems there isn't much code to be mutualized between the different functions. Maybe there is something I'm missing?

The mutualized code, that concerns the converter tests, is the initialization of the `input`, `output` and `ref` folders. These folders are common to all the converter tests.  Here, they are provided as arguments to the pytest call through the CI script (the `Jenkins` file in the root). It's true that for this case the mutualized code is not too much, but it allows to write a new test in only a few lines and (in the previous case, to create a new test we used to copy/paste all the function and the edit it). Moreover, if we plan to introduce multiprocessing during the tests, using fixtures can allow us to run the tests in parallel easily.

[the remark this answers is not in the document]
Comment (ghisvail) [reply test/nonregression/iotools/test_run_converters.py:52]: Then, I still don't understand what collapsing all the functions into a large one with branching brings to the table?

Looking at the diff, it does not sound to me that we are getting much benefits, while definitely losing in terms of readability and maintainability.

> In the previous case, to create a new test we used to copy/paste all the function and the edit it.

Separation of concerns is a good a thing. One converter one test function.

Anyway, I don't want to delay this review much longer. If @omar-rifai is happy with it, then let's go.

[the remark this answers is not in the document]
Comment (omar-rifai) [reply test/nonregression/iotools/test_run_converters.py:52]: > Moreover, if we plan to introduce multiprocessing during the tests, using fixtures can allow us to run the tests in parallel easily.

This kind of sells it for me. One thing we could potentially do to meet halfway is to keep a single entry point function and then have auxiliary functions for each test.

[the remark this answers is not in the document]
Comment (mdiazmel) [reply test/nonregression/iotools/test_run_converters.py:52]: It sounds like a good compromise! I will add single functions for each test as entry point and some other modifications.

[the remark this answers is not in the document]
Comment (mdiazmel) [reply test/nonregression/iotools/test_run_converters.py]: Hey @omar-rifai and @ghisvail ,

I rolled back to the idea of having single functions for the tests (called now `run_converter`) but keeping the fixture that wraps the main test function (and that allow us to have separate contexts for each test). What do you thing?

**Comment by omar-rifai** [comment]
Thanks @mdiazmel Maybe we can do the same (i.e entry point function) for the test utils as well? Also, (@ghisvail ?) for Nifd2Bids and Oasis32Bids the tests fail with an import error. For Nifd2Bids I updated my local data with the latest from data_ci in the lustre but couldn't find those for Oasis32Bids.

**Comment by ghisvail** [comment]
> [...] but couldn't find those for Oasis32Bids.

@MatthieuJoulot could you check this?

> [...] the tests fail with an import error.

Some logs would help.

**Comment by omar-rifai** [comment]
> > [...] but couldn't find those for Oasis32Bids.
> 
> @MatthieuJoulot could you check this?
> 
> > [...] the tests fail with an import error.
> 
> Some logs would help.
Sorry wrong error (it is rather a None Type error) 

`x = None

>       .rename_axis(index=lambda x: x.lower().replace(" ", "_"))
        .convert_dtypes()
        .sort_index()
    )
E   AttributeError: 'NoneType' object has no attribute 'lower'`
--> Looking into it

**Comment by ghisvail** [comment]
> > > [...] but couldn't find those for Oasis32Bids.
> > 
> > 
> > @MatthieuJoulot could you check this?
> > > [...] the tests fail with an import error.
> > 
> > 
> > Some logs would help.
> > Sorry wrong error (it is rather a None Type error)
> 
> > ` x = None         .rename_axis(index=lambda x: x.lower().replace(" ", "_"))         .convert_dtypes()         .sort_index()     ) E   AttributeError: 'NoneType' object has no attribute 'lower'`
> > --> Looking into it

On OASIS3 or NIFD? This is not clear from the logs.

**Comment by omar-rifai** [comment]
> > > > [...] but couldn't find those for Oasis32Bids.
> > > 
> > > 
> > > @MatthieuJoulot could you check this?
> > > > [...] the tests fail with an import error.
> > > 
> > > 
> > > Some logs would help.
> > > Sorry wrong error (it is rather a None Type error)
> > 
> > 
> > > ` x = None         .rename_axis(index=lambda x: x.lower().replace(" ", "_"))         .convert_dtypes()         .sort_index()     ) E   AttributeError: 'NoneType' object has no attribute 'lower'`
> > > --> Looking into it
> 
> On OASIS3 or NIFD? This is not clear from the logs.

NIFD

**Comment by ghisvail** [comment]
Fyi, for NIFD keep in mind we have postfixes waiting to be merged: #403 and #408

**Comment by omar-rifai** [comment]
> Fyi, for NIFD keep in mind we have postfixes waiting to be merged: #403 and #408

Ok, if you confirm this is the issue, then we can merge.

**Comment by ghisvail** [comment]
> Ok, if you confirm this is the issue, then we can merge.

All I can confirm is the respective issues being fixed by these merge requests.

Maybe what you are experiencing is something totally different though.

There is only one way to find out.

**Comment by mdiazmel** [comment]
> [...] Also for Nifd2Bids the tests fail with a None Type Error (I'm looking into it). 

It's certainly an error with your local data, the convertor works and [the tests are ok in the CI machines](https://ci.inria.fr/clinica-aramis/blue/organizations/jenkins/clinica/detail/PR-404/20/pipeline/92).

> [...] but couldn't find those for Oasis32Bids.

`Oasis32Bids` convertor is not concerned by this PR. If test will be added for this converter, I suppose that it'll be done in a different PR.

**Comment by ghisvail** [comment]
> the convertor works and the tests are ok in the CI machines.

Ok, great. That being said, the error is still worth investigating afterwards imo.

 > If test will be added for this converter, I suppose that it'll be done in a different PR.

Tests for OASIS3 should be added eventually (@MatthieuJoulot has got a test dataset).

I agree it should be done in a separate PR to avoid further delay on this one.

**Comment by omar-rifai** [comment]
> > [...] Also for Nifd2Bids the tests fail with a None Type Error (I'm looking into it).
> 
> It's certainly an error with your local data, the convertor works and [the tests are ok in the CI machines](https://ci.inria.fr/clinica-aramis/blue/organizations/jenkins/clinica/detail/PR-404/20/pipeline/92).


It seems the data in the lustre was not up-to-date. After retrieving the data directly from the CI machines, it seems to work.

**Review by omar-rifai — APPROVED** [review]

**Comment by ghisvail** [comment]
> It seems the data in the lustre was not up-to-date.

It might be worth redownloading the NIFD dataset entirely, since it seems LONI likes to change the imaging data paths as @omar-rifai observed with ADNI / AIBL last week. That being said, it would be even nicer if the converter could handle both the old and new formats in a reliable fashion. Maybe.
